File: packages/contracts/python/science_workbench_contracts/protocols/runtime.py
from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import Annotated, Final, Literal, Protocol, Self, override

from pydantic import Field, JsonValue, model_validator
from pydantic_core import PydanticCustomError

from science_workbench_contracts.common import NonEmptyText, Sha256, UtcTimestamp, Uuid7

from .approval import canonical_arguments_hash
from .models import Message, ProtocolModel, ToolResult

RuntimeErrorCategory = Literal[
    "quota", "reauth_required", "cancelled", "unavailable", "failed"
]
RuntimeTerminalOutcome = Literal["completed", "failed", "cancelled", "unavailable"]
RUNTIME_CORRELATION_ERROR: Final = "runtime_correlation"
RUNTIME_CORRELATION_MESSAGE: Final = "runtime command references do not correlate"


class RuntimeTextDelta(ProtocolModel):
    kind: Literal["text_delta"] = "text_delta"
    run_id: Uuid7
    message_id: Uuid7
    delta: NonEmptyText
    created_at: UtcTimestamp


class RuntimeActionProposal(ProtocolModel):
    kind: Literal["action_proposal"] = "action_proposal"
    run_id: Uuid7
    message_id: Uuid7
    action_id: Uuid7
    action_type: Literal["tool"] = "tool"
    tool: NonEmptyText
    arguments: JsonValue
    arguments_hash: Sha256
    continuation_token: NonEmptyText
    created_at: UtcTimestamp


class RuntimeTerminal(ProtocolModel):
    kind: Literal["terminal"] = "terminal"
    run_id: Uuid7
    outcome: RuntimeTerminalOutcome
    detail: NonEmptyText | None = None
    created_at: UtcTimestamp


class RuntimeQuota(ProtocolModel):
    kind: Literal["quota"] = "quota"
    run_id: Uuid7
    detail: NonEmptyText
    created_at: UtcTimestamp


class RuntimeReauthentication(ProtocolModel):
    kind: Literal["reauth_required"] = "reauth_required"
    run_id: Uuid7
    detail: NonEmptyText
    created_at: UtcTimestamp


RuntimeEvent = Annotated[
    RuntimeTextDelta
    | RuntimeActionProposal
    | RuntimeTerminal
    | RuntimeQuota
    | RuntimeReauthentication,
    Field(discriminator="kind"),
]


class RuntimeStart(ProtocolModel):
    run_id: Uuid7
    provider_connection_id: Uuid7
    message: Message

    @model_validator(mode="after")
    def correlated_message(self) -> Self:
        if self.message.run_id != self.run_id:
            raise PydanticCustomError(
                RUNTIME_CORRELATION_ERROR, RUNTIME_CORRELATION_MESSAGE
            )
        return self


class RuntimeContinue(ProtocolModel):
    org_id: Uuid7
    run_id: Uuid7
    message_id: Uuid7
    action_id: Uuid7
    execution_id: Uuid7
    continuation_token: NonEmptyText
    result: ToolResult

    @model_validator(mode="after")
    def correlated_result(self) -> Self:
        command_context = (
            self.org_id,
            self.run_id,
            self.message_id,
            self.action_id,
            self.execution_id,
        )
        result_context = (
            self.result.org_id,
            self.result.run_id,
            self.result.message_id,
            self.result.action_id,
            self.result.execution_id,
        )
        if command_context != result_context:
            raise PydanticCustomError(
                RUNTIME_CORRELATION_ERROR, RUNTIME_CORRELATION_MESSAGE
            )
        return self


class RuntimeCancel(ProtocolModel):
    run_id: Uuid7
    reason: NonEmptyText | None = None


class RuntimeErrorSignal(ProtocolModel):
    run_id: Uuid7
    category: RuntimeErrorCategory
    detail: NonEmptyText
    created_at: UtcTimestamp


@dataclass(frozen=True, slots=True)
class RuntimeProtocolError(Exception):
    code: Literal["ACTION_ARGUMENTS_HASH_MISMATCH"]

    @override
    def __str__(self) -> str:
        return self.code


class AgentRuntimeAdapter(Protocol):
    def start(self, command: RuntimeStart) -> AsyncIterator[RuntimeEvent]: ...

    async def continue_with_tool_result(self, command: RuntimeContinue) -> None: ...

    async def cancel(self, command: RuntimeCancel) -> None: ...


def validate_action_proposal(proposal: RuntimeActionProposal) -> None:
    if canonical_arguments_hash(proposal.arguments) != proposal.arguments_hash:
        raise RuntimeProtocolError(code="ACTION_ARGUMENTS_HASH_MISMATCH")


def normalize_runtime_error(signal: RuntimeErrorSignal) -> RuntimeEvent:
    event: dict[RuntimeErrorCategory, RuntimeEvent] = {
        "quota": RuntimeQuota(
            run_id=signal.run_id,
            detail=signal.detail,
            created_at=signal.created_at,
        ),
        "reauth_required": RuntimeReauthentication(
            run_id=signal.run_id,
            detail=signal.detail,
            created_at=signal.created_at,
        ),
        "cancelled": RuntimeTerminal(
            run_id=signal.run_id,
            outcome="cancelled",
            detail=signal.detail,
            created_at=signal.created_at,
        ),
        "unavailable": RuntimeTerminal(
            run_id=signal.run_id,
            outcome="unavailable",
            detail=signal.detail,
            created_at=signal.created_at,
        ),
        "failed": RuntimeTerminal(
            run_id=signal.run_id,
            outcome="failed",
            detail=signal.detail,
            created_at=signal.created_at,
        ),
    }
    return event[signal.category]

File: packages/contracts/src/protocols/runtime.ts
import { z } from "zod"

import { NonEmptyTextSchema, Sha256Schema, UtcTimestampSchema, Uuid7Schema } from "../common"
import { MessageSchema, ToolResultSchema } from "./models"

export const RuntimeTextDeltaSchema = z
  .strictObject({
    kind: z.literal("text_delta"),
    run_id: Uuid7Schema,
    message_id: Uuid7Schema,
    delta: NonEmptyTextSchema,
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const RuntimeActionProposalSchema = z
  .strictObject({
    kind: z.literal("action_proposal"),
    run_id: Uuid7Schema,
    message_id: Uuid7Schema,
    action_id: Uuid7Schema,
    action_type: z.literal("tool"),
    tool: NonEmptyTextSchema,
    arguments: z.json(),
    arguments_hash: Sha256Schema,
    continuation_token: NonEmptyTextSchema,
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const RuntimeTerminalSchema = z
  .strictObject({
    kind: z.literal("terminal"),
    run_id: Uuid7Schema,
    outcome: z.enum(["completed", "failed", "cancelled", "unavailable"]),
    detail: NonEmptyTextSchema.nullable().default(null),
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const RuntimeQuotaSchema = z
  .strictObject({
    kind: z.literal("quota"),
    run_id: Uuid7Schema,
    detail: NonEmptyTextSchema,
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const RuntimeReauthenticationSchema = z
  .strictObject({
    kind: z.literal("reauth_required"),
    run_id: Uuid7Schema,
    detail: NonEmptyTextSchema,
    created_at: UtcTimestampSchema,
  })
  .readonly()
export const RuntimeEventSchema = z.discriminatedUnion("kind", [
  RuntimeTextDeltaSchema,
  RuntimeActionProposalSchema,
  RuntimeTerminalSchema,
  RuntimeQuotaSchema,
  RuntimeReauthenticationSchema,
])
export const RuntimeStartSchema = z
  .strictObject({
    run_id: Uuid7Schema,
    provider_connection_id: Uuid7Schema,
    message: MessageSchema,
  })
  .superRefine((command, context) => {
    if (command.run_id !== command.message.run_id) {
      context.addIssue({ code: "custom", message: "runtime command references do not correlate" })
    }
  })
  .readonly()
export const RuntimeContinueSchema = z
  .strictObject({
    org_id: Uuid7Schema,
    run_id: Uuid7Schema,
    message_id: Uuid7Schema,
    action_id: Uuid7Schema,
    execution_id: Uuid7Schema,
    continuation_token: NonEmptyTextSchema,
    result: ToolResultSchema,
  })
  .superRefine((command, context) => {
    const result = command.result
    const correlated =
      command.org_id === result.org_id &&
      command.run_id === result.run_id &&
      command.message_id === result.message_id &&
      command.action_id === result.action_id &&
      command.execution_id === result.execution_id
    if (!correlated) {
      context.addIssue({ code: "custom", message: "runtime command references do not correlate" })
    }
  })
  .readonly()
export const RuntimeCancelSchema = z
  .strictObject({ run_id: Uuid7Schema, reason: NonEmptyTextSchema.nullable().default(null) })
  .readonly()

export interface AgentRuntimeAdapter {
  start(
    command: z.infer<typeof RuntimeStartSchema>,
  ): AsyncIterable<z.infer<typeof RuntimeEventSchema>>
  continueWithToolResult(command: z.infer<typeof RuntimeContinueSchema>): Promise<void>
  cancel(command: z.infer<typeof RuntimeCancelSchema>): Promise<void>
}

export type RuntimeEvent = z.infer<typeof RuntimeEventSchema>
